### PR TITLE
Handle Terminal Errors in the AzureCluster Controller

### LINKED ISSUE #1973: Handle `TerminalErrors` in the AzureCluster controller
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
Currently, the AzureMachine controller handles both transient and terminal errors, but the AzureCluster controller only handles transient errors. By adding support for terminal errors, we can return terminal errors in Reconcile() for services like subnets.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by sayantani11**
**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1973 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Handling Terminal Errors in AzureCluster
```

### CODE REVIEW

**Comment by sayantani11** [comment]
@CecileRobertMichon I wanted to understand the variable part in the error... where can I look for it?

File: controllers/azurecluster_controller.go [hunk 1]
```diff
@@ -221,9 +222,18 @@ func (acr *AzureClusterReconciler) reconcileNormal(ctx context.Context, clusterS
 	}
 
 	if err := acs.Reconcile(ctx); err != nil {
-		// Handle transient errors
+		// Handle terminal and transient errors
 		var reconcileError azure.ReconcileError
 		if errors.As(err, &reconcileError) {
+			if reconcileError.IsTerminal() {
```
Comment (CecileRobertMichon) [inline controllers/azurecluster_controller.go:228]: a couple of build errors still:

```
controllers/azurecluster_controller.go:230:76: clusterScope.Name undefined (type *scope.ClusterScope has no field or method Name)
controllers/azurecluster_controller.go:231:17: clusterScope.SetFailureReason undefined (type *scope.ClusterScope has no field or method SetFailureReason)
controllers/azurecluster_controller.go:232:17: clusterScope.SetFailureMessage undefined (type *scope.ClusterScope has no field or method SetFailureMessage)
controllers/azurecluster_controller.go:233:17: clusterScope.SetNotReady undefined (type *scope.ClusterScope has no field or method SetNotReady)
controllers/azurecluster_controller.go:234:17: clusterScope.SetVMState undefined (type *scope.ClusterScope has no field or method SetVMState)
```

Comment (CecileRobertMichon) [reply controllers/azurecluster_controller.go:228]: instead of `Name()` you can use `ClusterName`

Comment (CecileRobertMichon) [reply controllers/azurecluster_controller.go:228]: AzureClusters also don't have a `FailureMessage` or `SetFailureReason` in status so line 231 and 232 can be removed

Comment (sayantani11) [reply controllers/azurecluster_controller.go:228]: ok

**Pull request closed without merging**